### PR TITLE
Incompatible tools pop out of workstations on use

### PR DESCRIPTION
**Pull request by harakka**
#### Summary
Infrastructure "Pop out incompatible tools from workstations"

#### Purpose of change
Enable safer removal of tool types from workstations' allowed lists, by making unallowed tools pop out on workstation activation.

This is required so that Concrete mixer audit (#67047) works properly. Otherwise you will be able to activate any previously inserted mixers in workstations, spawning infinite concrete mixer vehicles.

This will also apply to any other future PRs where attached tools' properties are changed in a way that isn't compatible with station usage.

#### Describe the solution
VEH_TOOLS part check their tools on activation, and pop out any tools that are now incompatible.

#### Describe alternatives you've considered
Begging someone c++ competent to do this.

#### Testing
* Tested with the save and repro steps from #71108, with the concrete mixer audit PR applied. Concrete mixer item pops out and doesn't show up in the workshop menu any more.
* Created a test save ([Mooresville-one-with-all-tools.tar.gz](https://github.com/CleverRaven/Cataclysm-DDA/files/14735286/Mooresville-one-with-all-tools.tar.gz)) using https://github.com/CleverRaven/Cataclysm-DDA/releases/tag/cdda-experimental-2023-05-17-1227 which introduced concrete mixers into workstations. Added every single tool you can into both kinds of workstations. Tested this save with the same setup as the test above, concrete mixer pops out and isn't visible in the workshop menu.

#### Additional context
I'm an extreme c++ noob, please be kind and thorough when explaining to me what I'm doing wrong and how I should do this instead.

### CODE REVIEW

File: src/vehicle_use.cpp [hunk 1]
```diff
@@ -2303,9 +2351,9 @@ void vehicle::build_interact_menu( veh_menu &menu, const tripoint &p, bool with_
 
     const std::optional<vpart_reference> vp_toolstation = vp.avail_part_with_feature( "VEH_TOOLS" );
     if( vp_toolstation && vp_toolstation->info().toolkit_info ) {
-        const size_t vp_idx = vp_toolstation->part_index();
         const std::string vp_name = vp_toolstation->part().name( /* with_prefix = */ false );
 
+        const size_t vp_idx = vp_toolstation->part_index();
```
Comment (akrieger) [inline src/vehicle_use.cpp:2356]: Can revert this change?

File: src/vehicle_use.cpp [hunk 2]
```diff
@@ -2303,9 +2351,9 @@ void vehicle::build_interact_menu( veh_menu &menu, const tripoint &p, bool with_
 
     const std::optional<vpart_reference> vp_toolstation = vp.avail_part_with_feature( "VEH_TOOLS" );
```
Comment (akrieger) [inline src/vehicle_use.cpp:2352]:
```suggestion
```

**Review by akrieger — COMMENTED** [review]
Seems fine, my comments are minor. But idk vehicle code at all.

File: src/vehicle_use.cpp [hunk 3]
```diff
@@ -2118,6 +2118,54 @@ void vehicle::build_interact_menu( veh_menu &menu, const tripoint &p, bool with_
         } );
     }
 
+    // Remove attached tools that have become incompatible with workstation because of migration etc
+    if( vp.avail_part_with_feature( "VEH_TOOLS" ) && !vp.get_tools().empty() ) {
+        const std::optional<vpart_reference> vp_toolstation = vp.avail_part_with_feature( "VEH_TOOLS" );
+        if( vp_toolstation->info().toolkit_info ) {
```
[the remark this answers is not in the document]
Comment (harakka) [reply src/vehicle_use.cpp:2124]: We actually already have a `vp_toolstation` in this scope later in the file around line 2352, because that's where parts of this are copypasted from :grin: This check was originally located there, but we need it to precede the  line 2169-starting block because obsolete tools need to be cleared before the "use tool x" menu entries are built.

[the remark this answers is not in the document]
Comment (akrieger) [reply src/vehicle_use.cpp:2124]: So it's possible that by popping out all the tools, later on there won't be any more `avail_part_with_feature( "VEH_TOOLS" )`? Gotcha.

File: src/vehicle_use.cpp [hunk 4]
```diff
@@ -2118,6 +2118,54 @@ void vehicle::build_interact_menu( veh_menu &menu, const tripoint &p, bool with_
...
+            const std::set<itype_id> &allowed_tool_types = vp_toolstation->info().toolkit_info->allowed_types;
+
+            std::set<itype_id> builtin_tool_types;
+            for( const auto &[tool, _] : vp_toolstation->info().get_pseudo_tools() ) {
+                builtin_tool_types.insert( tool );
+            }
```
[the remark this answers is not in the document]
Comment (harakka) [reply src/vehicle_use.cpp]: Good justification, thanks. Didn't really think about how auto affects how it reads.

[the remark this answers is not in the document]
Comment (harakka) [reply src/vehicle_use.cpp]: Well I did it anyway.

File: src/vehicle_use.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (harakka) [reply src/vehicle_use.cpp]: No, VEH_TOOLS isn't provided by a tool in the vehicle part. I'm just bad at explaining :laughing:

The pattern used elsewhere in the file is that we introduce a vpart_reference variable of vp.avail_prt_with_feature("THE_THING_WE_WANT_TO_ADD_TO_THE_MENU"), and this is done later in the file for VEH_TOOLS where the actual menu entries for adding and removing tools from the workstation are created. But I think it's fine to diverge from that pattern here and pull the variable here already. I've done that in the latest commit.
